artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueueConsumersImpl.java

```diff
@@ -0,0 +1,124 @@
+import org.apache.activemq.artemis.utils.collections.ResettableIterator;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.Spliterator;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.Consumer;
+
+/**
+ * This class's purpose is to hold the consumers.
+ *
+ * CopyOnWriteArraySet is used as the underlying collection to the PriorityCollection, as it is concurrent safe,
+ * but also lock less for a read path, which is our HOT path.
+ * Also it was the underlying collection previously used in QueueImpl, before we abstracted it out to support priority consumers.
+ *
+ * There can only be one resettable iterable view,
+ * A new iterable view is created on modification, this is to keep the read HOT path performent, BUT
+ * the iterable view changes only after reset so changes in the underlying collection are only seen after a reset,
+ *
+ * All other iterators created by iterators() method are not reset-able and are created on delegating iterator().
+ *
+ * @param <T> The type this class may hold, this is generic as can be anything that extends PriorityAware,
+ *         but intent is this is the QueueImpl:ConsumerHolder.
+ */
+public class QueueConsumersImpl<T extends PriorityAware> implements QueueConsumers<T> {
+
+   private final PriorityCollection<T> consumers = new PriorityCollection<>(CopyOnWriteArraySet::new);
+   private final Collection<T> unmodifiableConsumers = Collections.unmodifiableCollection(consumers);
+   private final AtomicReferenceFieldUpdater<QueueConsumersImpl, ResettableIterator> changedIteratorFieldUpdater =  AtomicReferenceFieldUpdater.newUpdater(QueueConsumersImpl.class, ResettableIterator.class, "changedIterator");
+   private volatile ResettableIterator<T> changedIterator;
+   private ResettableIterator<T> currentIterator = consumers.resettableIterator();
+
+   @Override
+   public Set<Integer> getPriorites() {
+      return consumers.getPriorites();
+   }
+
+   @Override
+   public boolean hasNext() {
+      return currentIterator.hasNext();
+   }
+
+   @Override
+   public T next() {
+      return currentIterator.next();
+   }
+
+   @Override
+   public QueueConsumers<T> reset() {
+      if (changedIterator != null) {
+         currentIterator = changedIterator;
+         changedIteratorFieldUpdater.compareAndSet(this, changedIterator, null);
+      } else {
+         currentIterator.reset();
+      }
+      return this;
+   }
+
+   @Override
+   public boolean add(T t) {
+      boolean result = consumers.add(t);
+      if (result) {
+         changedIteratorFieldUpdater.set(this, consumers.resettableIterator());
+      }
+      return result;
+   }
+
+   @Override
+   public boolean remove(T t) {
+      boolean result = consumers.remove(t);
+      if (result) {
+         changedIteratorFieldUpdater.set(this, consumers.resettableIterator());
```

Comment:
`lazeSet` is enough for single-writer/single-threaded semantic

Reply:
Agreed. But this is not hot path, and if anything would allow in future for consumers to be concurrently added or removed without the existing bit sync blocks there is today in queueimpl